device-api/src/topology/hwloc.rs

```diff
@@ -0,0 +1,188 @@
+use std::ffi::CString;
+use std::os::raw::c_int;
+use std::ptr;
+
+use lazy_static::lazy_static;
+use regex::Regex;
+
+use crate::topology::hwloc_binding::*;
+use crate::{DeviceError, DeviceResult};
+
+pub(crate) struct HwlocTopology {
+    topology: hwloc_topology_t,
+}
+
+impl HwlocTopology {
+    pub(crate) fn new() -> Self {
+        Self {
+            topology: std::ptr::null_mut(),
+        }
+    }
+
+    pub(crate) fn init_topology(&mut self) -> DeviceResult<()> {
+        if unsafe { hwloc_topology_init(&mut self.topology) } == 0 {
+            Ok(())
+        } else {
+            Err(DeviceError::hwloc_error(
+                "couldn't initialize hwloc library",
+            ))
+        }
+    }
+
+    pub(crate) fn set_io_types_filter(&mut self, filter: hwloc_type_filter_e) -> DeviceResult<()> {
+        if unsafe { hwloc_topology_set_io_types_filter(self.topology, filter) } == 0 {
+            Ok(())
+        } else {
+            Err(DeviceError::hwloc_error("couldn't set filter"))
+        }
+    }
+
+    pub(crate) fn load_topology(&mut self) -> DeviceResult<()> {
+        if unsafe { hwloc_topology_load(self.topology) } == 0 {
+            Ok(())
+        } else {
+            Err(DeviceError::hwloc_error("couldn't load topology"))
+        }
+    }
+
+    pub(crate) fn set_topology_from_xml(&mut self, xmlpath: &str) -> DeviceResult<()> {
+        let xml_path_cstr = CString::new(xmlpath).unwrap();
+        if unsafe { hwloc_topology_set_xml(self.topology, xml_path_cstr.as_ptr()) } == 0 {
+            Ok(())
+        } else {
+            Err(DeviceError::hwloc_error("couldn't set topology from xml"))
+        }
+    }
+
+    pub(crate) fn get_common_ancestor_obj(
+        &self,
+        dev1bdf: &str,
+        dev2bdf: &str,
+    ) -> DeviceResult<hwloc_obj_t> {
+        let dev1_obj = unsafe { hwloc_get_pcidev_by_busidstring(self.topology, dev1bdf) };
+        if dev1_obj.is_null() {
+            return Err(DeviceError::hwloc_error(format!(
+                "couldn't find object with the bus id {dev1bdf}"
+            )));
+        }
+
+        let dev2_obj = unsafe { hwloc_get_pcidev_by_busidstring(self.topology, dev2bdf) };
+        if dev2_obj.is_null() {
+            return Err(DeviceError::hwloc_error(format!(
+                "couldn't find object with the bus id {dev2bdf}"
+            )));
+        }
+
+        let ancestor = unsafe { hwloc_get_common_ancestor_obj(dev1_obj, dev2_obj) };
+        if ancestor.is_null() {
+            return Err(DeviceError::hwloc_error(format!(
+                "couldn't find a common ancestor for objects {dev1bdf} and {dev2bdf}"
+            )));
+        }
+
+        Ok(ancestor)
+    }
+
+    pub(crate) fn destroy_topology(&mut self) {
+        if !self.topology.is_null() {
+            unsafe { hwloc_topology_destroy(self.topology) };
+            self.topology = std::ptr::null_mut();
+        }
+    }
+}
+
+pub unsafe fn hwloc_get_common_ancestor_obj(
```

Comment:
Is there a reason why the access specifier for this function is `pub`? Are there other places it's used?